feat(components): Add stepped variation of progress bar

jjwlee94 opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  For help creating your pull request, you can [use this tool](https://atlantis.getjobber.com/?path=/story/guides-pull-request-title-generator--page)

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - components-native
    - design
    - eslint
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

<!-- Why did you do what you did? -->

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Added

- <!-- new features -->

### Changed

- <!-- changes in existing functionality -->

### Deprecated

- <!-- soon-to-be removed features -->

### Removed

- <!-- now removed features -->

### Fixed

- <!-- for any bug fixes -->

### Security

- <!-- in case of vulnerabilities -->

## Testing

<!-- How to test your changes. -->

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

chris-at-jobber commented on packages/components/src/ProgressBar/ProgressBar.css:
```diff
@@ -20,6 +27,10 @@ progress.ProgressBar[value]::-webkit-progress-bar {
 }
 
+progress.SteppedProgressBar[value]::-webkit-progress-bar {
+  background-color: var(--color-greyBlue--lightest);
```
Both variations should use the same background color, so we don't need this extra declaration

jjwlee94 replied:
For context, we had adjusted the inactive state background colour to be a bit darker than what Atlantis version had because it was a bit harder to see in the stepped variation. If we want to maintain consistency between the two variations I can change this back!

`var(--color-surface--background)`:
<img width="920" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/30a90a29-26bf-451b-a01b-0a2dd7f0759c">

`var(--color-greyBlue--lightest)`:
<img width="920" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/c3a98d9c-6997-4426-8828-65e3ef400f7a">

chris-at-jobber replied:
Fair! There is a PR against our [design package](https://github.com/GetJobber/atlantis/pull/1545) that will update `surface--background` to map to `greyBlue--lightest` in the next few-ish days so let's use the existing design `surface--background` token and let it absorb the updates when that lands.

jjwlee94 replied:
Oh good to know! Sounds good :+1: Fixed in [`cb7b493`](https://github.com/GetJobber/atlantis/pull/1653/commits/cb7b493612510ae26de4f8e87bba85472c8c6e3b).

chris-at-jobber commented on packages/components/src/ProgressBar/ProgressBar.tsx:
```diff
@@ -19,15 +19,53 @@ interface ProgressBarProps {
+   * Set the variation to the stepped progress bar
+   * @default false
+   */
+  readonly stepped?: boolean;
```
Do we want this to be boolean? Referencing @irmalcol comment [here](https://getjobber.slack.com/archives/C03BJHV3PMG/p1700666198355529?thread_ts=1698257310.303899&cid=C03BJHV3PMG), it sounds like we have a bit more flexibility with this property if we treat it as a string 

Would something like 

```
  @default progress
  variation?: progress | stepped; 
```

work here?

jjwlee94 replied:
Oops, this slipped my mind. Agree that having this as a string would be this more flexible for potential additions in the future! Changed in [`bce3e13`](https://github.com/GetJobber/atlantis/pull/1653/commits/bce3e13933b1f7fe6343fa71eba7647e6a575b83).

chris-at-jobber reviewed (commented):
Design looks great! Just gave it a test with VoiceOver and we'll need to make some adjustments - I walk through it in a video [here](https://us02web.zoom.us/clips/share/rraO6L3qxh-syPQMRQvgtMK1F0lWapK1aJSrfEUA78dpeXM-ESrAW2UR4Y7CuZE_0bm1m6XbxooPo5pJOdzJ6tFwtg.3cZzYT20pGSTNrFZ).

TLDR the current approach tells VoiceOver that there are 5 separate progress bars, all either 100% or 0% complete. This is a confusing way to convey the information, so we need to find a way to communicate the total progress state to the user in a single element.

In the example with 3/5 filled, we currently read:

"progress 100%, progress 100%, progress 100%, progress 0%, progress 0%"

How it should read:

"progress, 60%" OR "progress, 3 out of 5"

darryltec commented on packages/components/src/ProgressBar/ProgressBar.tsx:
```diff
@@ -19,15 +19,53 @@ interface ProgressBarProps {
+          const ariaLabel = `progress, ${step} out of ${totalSteps}`;
+
+          return (
+            <progress
```
to chris' point about voiceover, this is gonna get read out separately and since this is a non-standard way of showing progress, i'd use a `progressbar` role https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/progressbar_role along with the other `aria`s to read out the correct information.

darryltec commented on packages/components/src/ProgressBar/ProgressBar.test.tsx:
```diff
@@ -16,4 +16,11 @@ describe("with props", () => {
+    const { container } = render(
+      <ProgressBar currentStep={2} totalSteps={3} variation={"stepped"} />,
+    );
+    expect(container).toMatchSnapshot();
```
let's avoid writing more snapshot tests here. I'd assert the specific things that the user would see.

- there should be 3 steps
- 2 out of the 3 are filled
- 1 out of 3 is not filled

jjwlee94 commented on packages/components/src/ProgressBar/ProgressBar.test.tsx:
```diff
@@ -16,4 +16,25 @@ describe("with props", () => {
   });
+
+  describe("stepped variation", () => {
+    it("renders correctly", () => {
```
I'm not sure if this is too generic; I could separate these out into separate assertions if that's preferred.

darryltec replied:
yeah, should separate them out.  Basically, the description should tell you what it's testing. That helps in understanding the test better. And also help you write them haha.

jjwlee94 replied:
Applied in [`ecc624d`](https://github.com/GetJobber/atlantis/pull/1653/commits/ecc624d5423292ffdb6c1dc9346b313145e15492)!

darryltec reviewed (commented):
last few things then this solid!

chris-at-jobber commented:
Just wanted to pop in with a fun wrinkle that @Kristen-Haring and I came across when updating the Figma component today!

Our `base` size is 16px high, and the `small` size is 4px high, but the intended design of this new bar is 8px high.

Our proposed solution is
- add a new `size` variant named `smaller` that uses the 4px height
- update `small` to use 8px

An added bonus point is that it maps to the `spacing` [variable values](https://atlantis.getjobber.com/?path=/docs/design-spacing--page#values) as well (base = 16, small = 8, smaller = 4)

It would be a _visually_ breaking change, but looking at usages in Jobber, a relatively straightforward fix to update the 6-ish existing usages of `small` to `smaller`.

This might be better-suited as a follow-up PR but wanted to make sure it was captured. Apologies for this entering the discussion later in the game than we'd have liked!

chris-at-jobber commented:
The last item in the stepped bar has a margin-right that will throw off alignment:
![image](https://github.com/GetJobber/atlantis/assets/39704901/fa03690f-169f-4161-9404-3a23e51d9f7a)

Can we use `gap` on the wrapper instead of `margin-right` on the child elements to keep the spacing only where it needs to be?

Kristen-Haring commented:
Hellllllo - just checking out the PR and noticed that the spacing for progress bar when we get really wide (images above) looks too tight. Is there a way to indicate a percentage-based spacing vs straight pixels to give it more breathing room at wider implementations?

jjwlee94 commented:
@chris-at-jobber I've adjusted size small to 8px in height and smaller to 4px in height as standardized. I wasn't sure what you meant by your comment below, so let me know if there's anything else required here!

> An added bonus point is that it maps to the spacing [variable values](https://atlantis.getjobber.com/?path=/docs/design-spacing--page#values) as well (base = 16, small = 8, smaller = 4)

I've also changed the styling to use `gap` on the wrapper instead of `margin-right`; what do we think of using `var(--space-base)` for the spacing? (Currently it's set to `var(--space-small)` @Kristen-Haring 

| Small | Base |
| ----- | ----- |
| <img width="1840" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/24123cf4-c696-4c3c-bee4-aa719ffe24d1"> | <img width="1840" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/76fecb4e-ba19-40c5-932a-a1e7b9cf1c46"> |
| <img width="1840" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/f7c73452-0e5b-4a49-9e78-f83a7f16f6e6"> | <img width="1840" alt="image" src="https://github.com/GetJobber/atlantis/assets/82672426/10334b58-2e6e-4915-bd7e-a0af38103b9d"> |

Also, what are your thoughts on a consistent transition speed between the two variations? The default variation uses `var(--timing-slow)` while the stepped variation uses `var(--timing-quick)`.

Kristen-Haring commented:
Thanks so much for the screens, @jjwlee94! After referring back to the current implementation on prod in our signup flow, I prefer the `var(--space-small)` so we can keep it as is if it has to be static.

The `quick` animation is my preference since that works well for the `stepped` variant - does it seem like a good clip for the base progress bar version? I trust you with that decision, Jeewon.

jjwlee94 commented on docs/components/ProgressBar/Web.stories.tsx:
Added in [`66d3356`](https://github.com/GetJobber/atlantis/pull/1653/commits/66d3356674f861ed8d73e944bdab8a66f92a6fec); let me know if what I have is :ok_hand:

jjwlee94 commented on packages/components/src/ProgressBar/ProgressBar.css:
```diff
@@ -31,3 +39,19 @@ progress.ProgressBar[value]::-moz-progress-bar {
+progress.SteppedProgressBar[value]::-moz-progress-bar {
+  border-radius: calc(var(--progress-bar--height) / 2);
+  background-color: var(--color-interactive);
+}
```
Cleaned up this file! We do have a difference in transition speed between the default vs. stepped variation (see [comment](https://github.com/GetJobber/atlantis/pull/1653#issuecomment-1828901849)); if your team has a preference on keeping this consistent, I can remove this.

darryltec reviewed (approved):
I am good with this! @chris-at-jobber is design good with this?

chris-at-jobber reviewed (commented):
Design approved! Thanks for handling that screen-reader functionality @jjwlee94 🙏
